### DISCUSSION TITLE
Parameters that effect memory and compute usage

### DISCUSSION THREAD

**Discussion by laynr** [TorchIO-project/torchio#646] [Q&A] [answered]
I am running out of memory running the following code.  Would anyone be willing to provide insight into what parameters effect compute resources the most when using torchio?  Specifically wondering about 'queue_length',  'samples_per_volume', 'num_workers', and 'batch_size'.

```python
    # dataset contains  500 subjects
    patch_size = (128, 128, 64)
    queue_length = 300
    samples_per_volume = 10
    sampler = tio.data.UniformSampler(patch_size)
    patches_queue = tio.Queue(
        dataset,
        queue_length,
        samples_per_volume,
        sampler,
        num_workers=4,
    )
    patches_loader = DataLoader(patches_queue, batch_size=16)    

    num_epochs = 2
    model = torch.nn.Identity()
    #model = torch.hub.load('fepegar/highresnet', 'highres3dnet', pretrained=True)
    for epoch_index in range(num_epochs):
        for patches_batch in patches_loader:
            inputs = patches_batch['image'][tio.DATA]
            targets = patches_batch['label']
            logits = model(inputs)
    
    torch.save(model.state_dict(), f'./saved_model')
```

**Comment by fepegar** [comment, marked as answer]
Hi, @laynr. It might also be related to the way you build your `dataset`. First, what is the output of `patches_queue.get_max_memory_pretty()` ([docs](https://torchio.readthedocs.io/data/patch_training.html#torchio.data.Queue.get_max_memory_pretty))? How much RAM do you have?

**Reply by laynr** [reply]
@fepegar I am building my dataset with `tio.ScalarImage(<path to DICOM folder>)` as shown below.  Is there a more economical way to do this, should I write objects to disk?:


```
# iterate thru folders containing  patients DICOM images
for patient in patients:
    # get label for patient 
    label = labels_df._get_value(int(patient), 'patient_label')
    
     # create subject object for each patient
     subject_object= tio.Subject(
          image=tio.ScalarImage(f'{train_dir}{patient}',),
          label=label,
      )
     # add subject object to subjects_list
    subjects_list.append(subject_object)
    
# normalize and resize scans
transforms = [
   tio.ToCanonical(),
   tio.Resample(1),
   tio.ZNormalization(masking_method=tio.ZNormalization.mean),
   tio.CropOrPad((128,128,64)),
    ]
    transform = tio.Compose(transforms)
    
# create subject datasets
subjects_dataset = tio.SubjectsDataset(subjects_list, transform=transform)
```


I am using kaggle: 

- patches_queue.get_max_memory_pretty() results in 1.2 GiB

- Most time kaggle errors with a `Timeout Exceeded` after 9 hours even with just 100 patients.

**Reply by fepegar** [reply]
Everything looks all right. Can you please paste the error traceback from Kaggle? Is it a memory issue or a timeout issue?

**Reply by laynr** [reply]
@fepegar Thank you so much for your responses.

Every time I use ```model = torch.hub.load('fepegar/highresnet', 'highres3dnet', pretrained=True)``` instead of ```model = torch.nn.Identity()``` in the code above, after a couple minutes it returns ```Your notebook tried to allocate more memory than is available. It has restarted.``` it gives no other error traceback.

![image](https://user-images.githubusercontent.com/775607/132100120-f2eeb929-81c6-4b30-a765-2eb684dca85b.png)


When I use  ```model = torch.nn.Identity()```   and with over 100 subjects it times out, with the error below.  But I just think I need to give it more time on a different platform, because with 42 samples it completes with no errors. Using GPU or TPU didn't seem to speed it up much.

![image](https://user-images.githubusercontent.com/775607/132100194-adcfdb50-3f0b-478d-bbe1-b3c825991766.png)

**Reply by fepegar** [reply]
In that case, I guess your memory issue is GPU-related, as highresnet is very big and your patches are as well, so this issue doesn't seem to be caused by TorchIO.

**Reply by romainVala** [reply]
I agree this is not a torchio issue, but still, it will be good, to have an estimation (within torchio) of what GPU and CPU memory will be. no ?
(I did not play with tio.Queue.get_max_memory_pretty(), may it  is already here, it is only cpu, and takes account of num_worker ?)

**Reply by fepegar** [reply]
I'm not sure I understand your question, @romainVala.

**Reply by romainVala** [reply]
get_max_memory, is only computing the CPU  mem needed for the queue
but this is only half of the story : One should also compute the memory needed by each worker.
(which then depend on the total volume size, and the number of patch per volume)

(Note that , GPU memory is very different since it is only dependent on the input size, and the model used)

**Reply by fepegar** [reply]
The memory needed by each worker is much harder to compute because it depends on how much memory is used internally by each transform.

**Reply by romainVala** [reply]
yes you are right I forgot the transforms 
may be an estimate without transform (min memory) could be of interest, to understand (and help to choose) `num_worker` or `sample_per_volume`

And then for each transform this should linked to the volume input size ... and one could start with a rough estimate (2 or 3 * input_size) depending on the transform.

Not sure if it is possible, but it is important. I get often issues with cpu memory on my GPU cluster, and I have no other choice, than try and test, to adapt those parameters
